Add subprocess waitress app test fixture.

T4rk1n opened the pull request:
Add `dash_subprocess` fixture, runs a dash app in a subprocess waitress-serve command.

T4rk1n commented:
@alexcjohnson 

`dash_app` combine importing an app from a file into the current running process and starting it using `start_dash` in a thread.

**dash_subprocess**

- Runs as external process via the `waitress-serve` command.
- wsgi
- knows exactly when the server is ready to receive requests.
- Immediate stop.

I made it specifically to test packaging component libraries made with the boilerplate to make sure they work when published. Basically going to package a build, install it, launch the `usage.py` with the subprocess and test the component works. Cannot do that from the same process.

**start_dash**

- Runs a threaded instance of a dash app via `Dash.run_server` (same process).
- development server
- time.sleep for server start detection, if the app is too long to initialize it may fail.
- lengthy stop with a request + time.sleep

All in all, I'd say a developer probably wants to runs as a subprocess unless he needs to count callback, then he'd use the threaded approach and an app instance created inside the test.

I think the naming could be better, especially now that there's two non-blocking way to run an app, `start_dash`->`dash_threaded`. I am also thinking about adding an option to `dash_app` to run in subprocess or threaded and default to subprocess. So `dash_app` would stay the main fixture for running a dash app from a file.

> but it looks to me as though you don't need it with dash_app either, right?

Right, both methods will call `selenium.get` when the server starts.

Thanks for having a look.

alexcjohnson commented:
> All in all, I'd say a developer probably wants to runs as a subprocess unless he needs to count callback, then he'd use the threaded approach and an app instance created inside the test.

Makes sense, thanks for the explanation!

> I am also thinking about adding an option to `dash_app` to run in subprocess or threaded and default to subprocess. So `dash_app` would stay the main fixture for running a dash app from a file.

So like `dash_app('test_apps/simple_app.py', subprocess=False)`? That definitely sounds better. Only concern is that they have different syntax - `'test_apps/simple_app.py'` vs `'test_apps.simple_app'` - are the two equivalent? ie is it possible to convert one to the other, such that a `dash_app` user only needs to know one syntax, or is there some difference in module resolution that would invalidate this?

> Right, both methods will call `selenium.get` when the server starts.

Great - so please 🔪 the [unnecessary line](https://github.com/T4rk1n/pytest-dash/blob/f6584864dc506d7e39a1280ec88a2044ed6dfc25/tests/test_integrations.py#L9) and replace it with some dummy code that does something (like callback counting) that you can only do with the threaded approach.

T4rk1n commented:
The dot notation is not the same, but both are paths so it is easy to convert.

alexcjohnson commented on pytest_dash/tools.py:
```diff
@@ -43,12 +75,8 @@ def start_dash(selenium):
-            stopper()
-            return 'stop'
+        if '/stop' not in app.server.view_functions:
+            app.server.add_url_rule('/stop', '/stop', _stop_server)
```
Could `/stop` ever exist and *not* be a command to stop the server? If so seems like we'd need to search for a random unused route.

alexcjohnson commented on pytest_dash/tools.py:
```diff
@@ -106,8 +121,75 @@ def dash_app(dash_from_file, start_dash):
+
+    def _sub(app_module):
+        server_path = '{}:app.server'.format(
+            '.'.join(os.path.split(app_module.replace('.py', '')))
```
Hmm, I'm unsure about this - for one, now `.py` is optional for `dash_subprocess` but required for `dash_app`, so if first write your test for `dash_subprocess` and you omit `.py` you can't just switch to `dash_app`. That one we could solve by adding `.py` to `dash_app` if it's missing, but second, is the argument really a path here? In `dash_app` presumably you could include `../` in the path, like if you had a demo app in a directory parallel to your `tests/` directory and you want to use that in the tests. That would clearly break `'.'.join(...)`, would there be a way to access that app using `waitress`?

But then we get a new version of the original question - is there any good reason to use `dash_app` anymore, or would you always use `dash_subprocess(app_module_str)` for testing an app as a black box and always use `dash_threaded(app_object)` if you want to inspect the back end?

alexcjohnson commented:
Love it, you've covered everything I can see! 💃

T4rk1n commented:
There's only one thing left to do, make sure it works cross-platform, I test on windows and it works but on linux it doesn't.

T4rk1n commented:
Releasing this as 1.0.0 as I removed a fixture and renamed one. The subprocess fixture doesn't seem to work on circleci, but it worked on linux/windows/mac. I think it may be a path issue with how I tried to run the test on `dash-component-boilerplate` that may get in the way. I'll set up circle later on this repo so I can test that more efficiently.
